### PR TITLE
Migrate fullchain IT wih 1 badworker in the array

### PR DESCRIPTION
**Pull request by gfournieriExec**

### CODE REVIEW

File: test/000_fullchain.test.ts [hunk 1]
```diff
@@ -299,6 +300,135 @@ describe('Integration tests', function () {
                 for (let i = 0; i < workerNumber; i++) {
                     if (workerNumber == 1) {
                         expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(0);
+                    } else {
```
Comment (jeremyjams) [inline test/000_fullchain.test.ts:303]: Maybe `expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(workerNumber == 1? 0 : 1);
`

File: test/000_fullchain.test.ts [hunk 2]
```diff
@@ -352,41 +359,29 @@ describe('Integration tests', function () {
                     let frozen = (await iexecPoco.frozenOf(account.address)).toNumber();
                     accountsInitBalances.push({ address, frozen });
                 }
-                for (let i = 0; i < workerNumber; i++) {
-                    expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(0);
+                // for (let i = 0; i < workerNumber; i++) {
```
Comment (gfournieriExec) [inline test/000_fullchain.test.ts:362]: to be removed

File: test/000_fullchain.test.ts (hunk 1 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: updated [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/5c6045a181b7d5a60e17355026900ac504cfffaa)

File: test/000_fullchain.test.ts [hunk 3]
```diff
@@ -297,10 +298,126 @@ describe('Integration tests', function () {
...
+                        [winningWorker.address],
+                        [workerStake + workerRewardPerTask / goodWorkersNumber],
+                    );
+                    expect(await iexecPoco.viewScore(winningWorker.address)).to.be.equal(1);
+                }
+                // verify task status
+                expect((await iexecPoco.viewTask(taskId)).status).to.equal(
+                    TaskStatusEnum.COMPLETED,
+                );
+                // checks n frozen balance changes
```
Comment (gfournieriExec) [inline test/000_fullchain.test.ts:412]:
```suggestion
                // checks on frozen balance changes
```

**Review by jeremyjams — APPROVED** [review]

File: test/000_fullchain.test.ts [hunk 4]
```diff
@@ -297,10 +298,126 @@ describe('Integration tests', function () {
...
+                ];
+                // Create deal.
+                const orders = buildOrders({
+                    assets: ordersAssets,
+                    prices: ordersPrices,
+                    requester: requester.address,
+                    tag: standardDealTag,
+                    beneficiary: beneficiary.address,
+                    volume,
+                    trust: goodWorkersNumber,
```
[the remark this answers is not in the document]
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: done

File: test/000_fullchain.test.ts [hunk 5]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
-                    }
+                    expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(
+                        workerNumber == 1 ? 0 : 1,
+                    );
+                }
+            });
+        }
+        for (let workerNumber = 4; workerNumber < 6; workerNumber++) {
+            // Worker1 will contribute badly
+            it(`[7.${workerNumber - 3}] No sponsorship, no beneficiary, no callback, no BoT, up to ${workerNumber} workers with 1 bad actor`, async function () {
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:309]:
```suggestion
            it(`[7.${workerNumber - 3}] No sponsorship, no beneficiary, no callback, no BoT, up to ${workerNumber} workers with 1 bad worker`, async function () {
```

File: test/000_fullchain.test.ts [hunk 6]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
+                    await iexecWrapper.signAndMatchOrders(...orders.toArray());
+                const taskPrice = appPrice + datasetPrice + workerpoolPrice;
+                const schedulerStakePerTask = schedulerStakePerDeal / volume;
+                const workerRewardPerTask = await iexecWrapper.computeWorkerRewardPerTask(
+                    dealId,
+                    PocoMode.CLASSIC,
+                );
+                const schedulerRewardPerTask = workerpoolPrice - workerRewardPerTask;
+                // Check initial balances.
+                let accountsInitBalances = [
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:348]:
```suggestion
                let accountsInitialFrozens = [
```

File: test/000_fullchain.test.ts [hunk 7]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
+                    {
+                        address: proxyAddress,
+                        frozen: (await iexecPoco.frozenOf(proxyAddress)).toNumber(),
+                    },
+                ];
+                for (const account of acounts) {
+                    let address = account.address;
+                    let frozen = (await iexecPoco.frozenOf(account.address)).toNumber();
+                    accountsInitBalances.push({ address, frozen });
+                }
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:358]:
```suggestion
                for (const account of acounts) {
                    accountsInitialFrozens.push({
                        address: account.address,
                        frozen: (await iexecPoco.frozenOf(account.address)).toNumber(),
                    });
                }
```

Comment (zguesmi) [reply test/000_fullchain.test.ts:358]: Also you can put all of the above in a function `const accountsInitialFrozens = getInitialFrozens(accounts)` and use it in the other test.

File: test/000_fullchain.test.ts [hunk 8]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
+                for (const worker of workers) {
+                    expect(await iexecPoco.viewScore(worker.signer.address)).to.be.equal(0);
                 }
+                const taskId = await iexecWrapper.initializeTask(dealId, 0);
+                // Finalize each task and check balance changes.
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:363]:
```suggestion
                // Finalize task and check balance changes.
```

File: test/000_fullchain.test.ts [hunk 9]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
+            it(`[7.${workerNumber - 3}] No sponsorship, no beneficiary, no callback, no BoT, up to ${workerNumber} workers with 1 bad actor`, async function () {
+                const volume = 1;
+                const workersAvailable = [worker1, worker2, worker3, worker4, worker5];
+                const { resultDigest: badResultDigest } = buildUtf8ResultAndDigest('bad-result');
+                const losingWorker = worker1;
+                const winningWorkers = workersAvailable.slice(1, workerNumber);
+                let workers = [{ signer: worker1, resultDigest: badResultDigest }];
+                for (const worker of winningWorkers) {
+                    workers.push({ signer: worker, resultDigest: resultDigest });
+                }
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:318]: It seems to me more like `contributions` not `workers`
```suggestion
                for (const worker of winningWorkers) {
                    contributions.push({ signer: worker, resultDigest: resultDigest });
                }
```

File: test/000_fullchain.test.ts [hunk 10]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
+                    .then((deal) => deal.workerStake.toNumber());
+                for (const worker of workers) {
+                    await iexecWrapper.contributeToTask(
+                        dealId,
+                        0,
+                        worker.resultDigest,
+                        worker.signer,
+                    );
+                }
+                for (const winningWorker of winningWorkers) {
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:375]: You can also reveal for the losing worker and check that it fails.

File: test/000_fullchain.test.ts [hunk 11]
```diff
@@ -297,10 +298,125 @@ describe('Integration tests', function () {
...
-                    if (workerNumber == 1) {
-                        expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(0);
-                    }
+                    expect(await iexecPoco.viewScore(workers[i].address)).to.be.equal(
+                        workerNumber == 1 ? 0 : 1,
+                    );
+                }
+            });
+        }
+        for (let workerNumber = 4; workerNumber < 6; workerNumber++) {
```
Comment (zguesmi) [inline test/000_fullchain.test.ts:307]: I think we can keep only 1 test with either 4 or 5 workers and 1 bad worker. We can even do 5 workers with 2 bad workers. But I don't see the benefit of doing more.

File: test/000_fullchain.test.ts (hunk 10 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts:359]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/0764ba1e90e606906a52a302d8cf30cf1aba6072)

File: test/000_fullchain.test.ts (hunk 9 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/715c453c77ebbf3915c88e780564b53129cf4d78)

File: test/000_fullchain.test.ts (hunk 8 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/83a8e64ddaffed58c766134199b94c27165f98be)

File: test/000_fullchain.test.ts (hunk 7 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/83a8e64ddaffed58c766134199b94c27165f98be)

File: test/000_fullchain.test.ts (hunk 5 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: Done [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/83a8e64ddaffed58c766134199b94c27165f98be)

File: test/000_fullchain.test.ts (hunk 11 above)
Comment (gfournieriExec) [reply test/000_fullchain.test.ts:297]: Okay I will keep 1 test => 4 good 1 bad

Comment (gfournieriExec) [reply test/000_fullchain.test.ts]: Updated [here](https://github.com/iExecBlockchainComputing/PoCo/pull/160/commits/982f21c5ce13fddd16df334507bd33eef881b229)

**Review by zguesmi — APPROVED** [review]

**Pull request merged**